File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        List<Integer>[] graph = new ArrayList[n + 1];
        boolean[] visited = new boolean[n + 1];

        for (int i = 1; i < n + 1; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());

            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            graph[a].add(b);
            graph[b].add(a);
        }

        int count = 0;
        for (int i = 1; i < n + 1; i++) {
            if (!visited[i]) {
                count++;

                Queue<Integer> queue = new LinkedList<>();
                queue.add(i);
                visited[i] = true;

                while (!queue.isEmpty()) {
                    int cur = queue.poll();

                    for (int next : graph[cur]) {
                        if (!visited[next]) {
                            visited[next] = true;
                            queue.add(next);
                        }
                    }
                }
            }
        }

        System.out.println(count);
    }
}

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.kt
fun main() {
    val (n, m) = readln().split(" ").map(String::toInt)
    val graph = Array(n + 1) { mutableListOf<Int>() }
    val visited = BooleanArray(n + 1)
    repeat(m) {
        val (a, b) = readln().split(" ").map(String::toInt)

        graph[a].add(b)
        graph[b].add(a)
    }

    var count = 0
    for (i in 1..n) {

        if (!visited[i]) {
            val queue = ArrayDeque<Int>()
            queue.addLast(i)
            visited[i] = true
            count++

            while (queue.isNotEmpty()) {
                val cur = queue.removeFirst()

                for (next in graph[cur]) {
                    if (!visited[next]) {
                        visited[next] = true
                        queue.addLast(next)
                    }
                }
            }
        }
    }

    print(count)
}